File: apps/ios/Vostok/App/AppContainer.swift
import Foundation

final class AppContainer: ObservableObject {
    let environment: AppEnvironment
    let cryptoProvider: CryptoProviderProtocol
    let serverStore: ServerStore

    /// All configured servers and their service contexts.
    @Published private(set) var serverContexts: [String: ServerContext] = [:]
    /// The currently active server ID.
    @Published var activeServerId: String?

    /// The active server's context (convenience for feature views).
    var activeContext: ServerContext? {
        guard let id = activeServerId else { return nil }
        return serverContexts[id]
    }

    // MARK: – Legacy compatibility (delegate to active server context)

    var apiClient: VostokAPIClientProtocol { activeContext?.apiClient ?? fallbackAPIClient }
    var realtimeClient: PhoenixRealtimeClientProtocol { activeContext?.realtimeClient ?? fallbackRealtimeClient }
    var chatRepository: ChatRepository { activeContext?.chatRepository ?? fallbackChatRepo }
    var messageRepository: MessageRepository { activeContext?.messageRepository ?? fallbackMsgRepo }
    var callRepository: CallRepository { activeContext?.callRepository ?? fallbackCallRepo }
    var mediaRepository: MediaRepository { activeContext?.mediaRepository ?? fallbackMediaRepo }
    var mediaTransferService: MediaTransferService { activeContext?.mediaTransferService ?? fallbackMediaTransfer }
    var database: VostokDatabase { activeContext?.database ?? fallbackDatabase }
    var signalSessionRuntime: SignalSessionRuntimeProtocol { activeContext?.signalSessionRuntime ?? fallbackSessionRuntime }

    // MARK: – Fallbacks (used when no server is active, e.g. during onboarding)
    private let fallbackAPIClient: VostokAPIClientProtocol
    private let fallbackRealtimeClient: PhoenixRealtimeClientProtocol
    private let fallbackDatabase: VostokDatabase
    private let fallbackChatRepo: ChatRepository
    private let fallbackMsgRepo: MessageRepository
    private let fallbackCallRepo: CallRepository
    private let fallbackMediaRepo: MediaRepository
    private let fallbackMediaTransfer: MediaTransferService
    private let fallbackSessionRuntime: SignalSessionRuntimeProtocol

    init(environment: AppEnvironment) {
        self.environment = environment
        self.cryptoProvider = SignalCryptoProvider()
        self.serverStore = ServerStore()

        // Create fallback services using the Environment.plist URL
        let api = APIClient(baseURL: environment.baseURL)
        let realtime = PhoenixRealtimeClient(socketURL: environment.socketURL)
        let database: VostokDatabase
        do {
            database = try VostokDatabase()
        } catch {
            fatalError("Failed to initialize fallback SQLCipher database: \(error.localizedDescription)")
        }

        self.fallbackAPIClient = api
        self.fallbackRealtimeClient = realtime
        self.fallbackDatabase = database
        self.fallbackChatRepo = InMemoryChatRepository(apiClient: api, database: database)
        self.fallbackMsgRepo = InMemoryMessageRepository(apiClient: api, database: database)
        self.fallbackCallRepo = NetworkCallRepository(apiClient: api)
        self.fallbackMediaRepo = NetworkMediaRepository(apiClient: api)
        self.fallbackMediaTransfer = MediaTransferService(repository: NetworkMediaRepository(apiClient: api))
        self.fallbackSessionRuntime = SignalSessionRuntime(apiClient: api, database: database)
    }

    // MARK: – Server lifecycle

    /// Loads servers from disk and creates contexts for each.
    /// Migrates from legacy single-server if needed.
    func loadServers() {
        var servers = serverStore.load()

        // Migrate from single-server if this is the first multi-server launch
        if servers.isEmpty {
            servers = serverStore.migrateFromLegacySingleServer(environment: environment)
        }

        for server in servers where server.enabled {
            if serverContexts[server.id] == nil {
                createContext(for: server)
            }
        }

        // Set active server to first one if none selected
        if activeServerId == nil, let first = servers.first(where: { $0.enabled }) {
            activeServerId = first.id
        }
    }

    /// Creates and registers a ServerContext for the given entry.
    @discardableResult
    func createContext(for entry: ServerEntry) -> ServerContext? {
        do {
            let context = try ServerContext(entry: entry)
            serverContexts[entry.id] = context
            return context
        } catch {
            print("[MultiServer] Failed to create context for \(entry.label): \(error)")
            return nil
        }
    }

    /// Removes a server and its context, database, and credentials.
    func removeServer(id: String) {
        if let context = serverContexts[id] {
            Task { await context.realtimeClient.disconnect() }
        }
        serverContexts.removeValue(forKey: id)
        _ = serverStore.remove(id: id)
        ServerDatabaseManager.removeDatabase(for: id)

        if activeServerId == id {
            activeServerId = serverStore.load().first(where: { $0.enabled })?.id
        }
    }

    /// Returns the context for a specific server, or nil if not loaded.
    func context(for serverId: String) -> ServerContext? {
        serverContexts[serverId]
    }

    /// Returns the context that owns a qualified chat ID.
    func context(forQualifiedChatId qualifiedId: String) -> ServerContext? {
        guard let serverId = QualifiedChatIdParser.serverId(from: qualifiedId) else {
            return activeContext
        }
        return serverContexts[serverId] ?? activeContext
    }

    /// Connects all enabled servers (establishes WebSocket connections).
    func connectAll(getToken: (String) -> String?) async {
        for (serverId, context) in serverContexts {
            guard let token = getToken(serverId) else { continue }
            context.connectionStatus = .connecting
            await context.realtimeClient.connect(token: token)
            context.connectionStatus = .connected
        }
    }

    /// Disconnects all server WebSocket connections.
    func disconnectAll() async {
        for (_, context) in serverContexts {
            await context.realtimeClient.disconnect()
            context.connectionStatus = .disconnected
        }
    }

    /// Returns merged and sorted chats from all connected servers.
    func allMergedChats() -> [MergedChatSummary] {
        var all: [MergedChatSummary] = []
        let servers = serverStore.load()

        for server in servers where server.enabled {
            guard let context = serverContexts[server.id] else { continue }
            let chats = context.database.loadChats()
            all.append(contentsOf: chats.map { MergedChatSummary.from(server: server, chat: $0) })
        }

        return sortMergedChats(all)
    }
}

File: apps/android/app/src/main/java/chat/vostok/android/app/AppContainer.kt
package chat.vostok.android.app

import android.content.Context
import chat.vostok.android.BuildConfig
import chat.vostok.android.core.crypto.KeyManager
import chat.vostok.android.core.crypto.SignalSessionRuntime
import chat.vostok.android.core.multiserver.ServerConnectionStatus
import chat.vostok.android.core.multiserver.ServerContext
import chat.vostok.android.core.multiserver.ServerDatabaseManager
import chat.vostok.android.core.multiserver.ServerEntry
import chat.vostok.android.core.multiserver.ServerStore
import chat.vostok.android.core.multiserver.serverIdFrom
import chat.vostok.android.core.network.ApiClient
import chat.vostok.android.core.network.WebSocketManager
import chat.vostok.android.core.repository.AuthRepository
import chat.vostok.android.core.repository.CallRepository
import chat.vostok.android.core.repository.ChatRepository
import chat.vostok.android.core.repository.ContactRepository
import chat.vostok.android.core.repository.DeviceRepository
import chat.vostok.android.core.repository.GroupRepository
import chat.vostok.android.core.repository.MessageRepository
import chat.vostok.android.core.repository.MediaRepository
import chat.vostok.android.core.storage.MediaCache
import chat.vostok.android.core.storage.SecurePreferencesFactory
import chat.vostok.android.core.storage.SessionStore
import chat.vostok.android.core.storage.VostokDatabase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.security.MessageDigest
import java.util.Base64

class AppContainer(private val context: Context) {
    private val tokenProvider = SessionTokenProvider()

    val keyManager = KeyManager(context)
    val sessionStore = SessionStore(context)
    val serverStore = ServerStore(context)
    val secureStorageStatus = SecurePreferencesFactory.currentStatus(context)
    val signingStorageSummary = keyManager.signingStorageSummary()

    // MARK: – Multi-server state
    private val _serverContexts = mutableMapOf<String, ServerContext>()
    val serverContexts: Map<String, ServerContext> get() = _serverContexts

    private val _activeServerId = MutableStateFlow<String?>(null)
    val activeServerId: StateFlow<String?> = _activeServerId.asStateFlow()

    val activeContext: ServerContext?
        get() = _activeServerId.value?.let { _serverContexts[it] }

    // MARK: – Legacy compatibility (delegates to active server context or fallback)
    private val fallbackApiClient: ApiClient
    private val fallbackWebSocket: WebSocketManager
    private val fallbackDatabase: VostokDatabase
    private val fallbackAuthRepository: AuthRepository
    private val fallbackChatRepository: ChatRepository
    private val fallbackMessageRepository: MessageRepository
    private val fallbackContactRepository: ContactRepository
    private val fallbackDeviceRepository: DeviceRepository
    private val fallbackGroupRepository: GroupRepository
    private val fallbackCallRepository: CallRepository
    private val fallbackMediaRepository: MediaRepository
    private val fallbackSignalRuntime: SignalSessionRuntime

    val apiClient: ApiClient get() = activeContext?.apiClient ?: fallbackApiClient
    val webSocketManager: WebSocketManager get() = activeContext?.webSocketManager ?: fallbackWebSocket
    val database: VostokDatabase get() = activeContext?.database ?: fallbackDatabase
    val authRepository: AuthRepository get() = activeContext?.authRepository ?: fallbackAuthRepository
    val chatRepository: ChatRepository get() = activeContext?.chatRepository ?: fallbackChatRepository
    val messageRepository: MessageRepository get() = activeContext?.messageRepository ?: fallbackMessageRepository
    val contactRepository: ContactRepository get() = activeContext?.contactRepository ?: fallbackContactRepository
    val deviceRepository: DeviceRepository get() = activeContext?.deviceRepository ?: fallbackDeviceRepository
    val groupRepository: GroupRepository get() = activeContext?.groupRepository ?: fallbackGroupRepository
    val callRepository: CallRepository get() = activeContext?.callRepository ?: fallbackCallRepository
    val mediaRepository: MediaRepository get() = activeContext?.mediaRepository ?: fallbackMediaRepository
    val signalSessionRuntime: SignalSessionRuntime get() = activeContext?.signalSessionRuntime ?: fallbackSignalRuntime
    val mediaCache = MediaCache(context)

    init {
        // Create fallback services from BuildConfig (legacy single-server)
        fallbackApiClient = ApiClient(baseUrl = BuildConfig.VOSTOK_BASE_URL) { tokenProvider.get() }
        fallbackWebSocket = WebSocketManager(socketUrl = BuildConfig.VOSTOK_SOCKET_URL)
        fallbackDatabase = VostokDatabase.getInstance(context, deriveDatabasePassphrase())

        fallbackSignalRuntime = SignalSessionRuntime(
            apiClient = fallbackApiClient,
            signalSessionDao = fallbackDatabase.signalSessionDao(),
            keyManager = keyManager,
            sessionStore = sessionStore,
        )
        fallbackAuthRepository = AuthRepository(
            apiClient = fallbackApiClient,
            keyManager = keyManager,
            sessionStore = sessionStore,
            tokenSetter = tokenProvider::set,
        )
        fallbackChatRepository = ChatRepository(
            apiClient = fallbackApiClient,
            chatDao = fallbackDatabase.chatDao(),
            messageDao = fallbackDatabase.messageDao(),
        )
        fallbackMessageRepository = MessageRepository(
            apiClient = fallbackApiClient,
            messageDao = fallbackDatabase.messageDao(),
            pendingOutboxDao = fallbackDatabase.pendingOutboxDao(),
            sessionRuntime = fallbackSignalRuntime,
        )
        fallbackContactRepository = ContactRepository(apiClient = fallbackApiClient)
        fallbackDeviceRepository = DeviceRepository(apiClient = fallbackApiClient)
        fallbackGroupRepository = GroupRepository(apiClient = fallbackApiClient)
        fallbackCallRepository = CallRepository(apiClient = fallbackApiClient)
        fallbackMediaRepository = MediaRepository(apiClient = fallbackApiClient, mediaCache = mediaCache)

        tokenProvider.set(sessionStore.load()?.token)
    }

    // MARK: – Server lifecycle

    fun loadServers() {
        var servers = serverStore.load()
        if (servers.isEmpty()) {
            servers = serverStore.migrateFromLegacySingleServer(sessionStore)
        }

        for (server in servers.filter { it.enabled }) {
            if (server.id !in _serverContexts) {
                createContext(server)
            }
        }

        if (_activeServerId.value == null) {
            _activeServerId.value = servers.firstOrNull { it.enabled }?.id
        }
    }

    fun createContext(entry: ServerEntry): ServerContext? {
        return try {
            val ctx = ServerContext(entry, context, keyManager, sessionStore)
            _serverContexts[entry.id] = ctx
            ctx
        } catch (e: Exception) {
            android.util.Log.e("AppContainer", "Failed to create context for ${entry.label}: $e")
            null
        }
    }

    fun removeServer(id: String) {
        _serverContexts.remove(id)
        serverStore.remove(id)
        ServerDatabaseManager.removeDatabaseFiles(context, id)
        if (_activeServerId.value == id) {
            _activeServerId.value = serverStore.load().firstOrNull { it.enabled }?.id
        }
    }

    fun contextForServer(serverId: String): ServerContext? = _serverContexts[serverId]

    fun contextForChat(qualifiedChatId: String): ServerContext? {
        val serverId = serverIdFrom(qualifiedChatId) ?: return activeContext
        return _serverContexts[serverId] ?: activeContext
    }

    fun setActiveServer(serverId: String) {
        _activeServerId.value = serverId
    }

    private fun deriveDatabasePassphrase(): ByteArray {
        val identityPublicKeyBase64 = keyManager.publicKeyBase64()
        return MessageDigest.getInstance("SHA-256")
            .digest(Base64.getDecoder().decode(identityPublicKeyBase64))
    }
}
